### PR TITLE
Tasks for Tn1

### PR DESCRIPTION
**Pull request by tudor-malene**

### CODE REVIEW

File: Todo TN1.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    - from the master seed derive keys. One public/private for the cluster and one symmetric per block
+    - implement viewing key support
+    - encrypt off-chain responses with sender viewing key
+    - encrypt transactions in rollup with generated symmetric key
+    - reveal key after few blocks
+    - tooling to decrypt (integrated with joel's block explorer)
+    - encrypt shared secret with a key that is derived from one of the burnt in sgx keys
+
+
+- Create better logic to get L2 node up to speed. (Very primitive logic in place now, not fit for purpose if connected to ropsten)
```
Comment (GavT) [inline Todo TN1.md:22]: This is T5

File: Todo TN1.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+#Obscuro:
+
+- evm /statedb - finalise & test properly (make sure it works through reorgs)
+
+- k/v on top of edgelessdb
+
+- add support for multiple supported erc20s (mapping L1/L2 address + withdrawals)
+
+- Implement the obscuro cryptography:
```
Comment (GavT) [inline Todo TN1.md:12]: This is T8

File: Todo TN1.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
+Todo TN1
+---------
+
+#Obscuro:
+
+- evm /statedb - finalise & test properly (make sure it works through reorgs)
+
+- k/v on top of edgelessdb
+
+- add support for multiple supported erc20s (mapping L1/L2 address + withdrawals)
```
Comment (otherview) [inline Todo TN1.md:10]: I thought this was done, is there stuff missing

File: Todo TN1.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+- implement better error handling - not perfect, but at least remove panics in places which are likely to blow up in a testnet scenario 
+
+- implment a better way for vending the shared secret (not everyone should respond)
+
+- some restructuring of the repo (renaming of packages, consistency, etc) + better documentation, to make it easier for reviewers to understand what's gong on
+
+
+# Management contract:
+
+Implement a primitive management contract that:
```
Comment (GavT) [inline Todo TN1.md:42]: Management contract or rollup contract? This needs to be a new Epic BTW

File: Todo TN1.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    - encrypt off-chain responses with sender viewing key
+    - encrypt transactions in rollup with generated symmetric key
+    - reveal key after few blocks
+    - tooling to decrypt (integrated with joel's block explorer)
+    - encrypt shared secret with a key that is derived from one of the burnt in sgx keys
+
+
+- Create better logic to get L2 node up to speed. (Very primitive logic in place now, not fit for purpose if connected to ropsten)
+
+- some memory analysis - make sure a node can stay up for at least a day of rollups
```
Comment (otherview) [inline Todo TN1.md:24]: should we not write to disk ?

File: Todo TN1.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+- rpc endpoints for everything necessary 
+- introduce event feeds (are these necessary for tn1?)
+
+- add configuration file support
```
Comment (GavT) [inline Todo TN1.md:29]: Captured as a ticket but not an Epic: https://github.com/obscuronet/obscuro-internal/issues/335
Should it be an Epic?

File: Todo TN1.md (hunk 3 above)
Comment (tudor-malene) [reply Todo TN1.md:10]: we need the mapping from L1 erc20 address to L2 erc20 address and we need to add it to the withdrawal instruction

File: Todo TN1.md (hunk 2 above)
Comment (tudor-malene) [reply Todo TN1.md:12]: yeah, but with more granularity

File: Todo TN1.md (hunk 1 above)
Comment (tudor-malene) [reply Todo TN1.md:22]: yes

File: Todo TN1.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (tudor-malene) [reply Todo TN1.md:24]: no, something different

Comment (tudor-malene) [reply Todo TN1.md:24]: yeah, after we do that, we need to do some analysys

File: Todo TN1.md (hunk 6 above)
Comment (tudor-malene) [reply Todo TN1.md:29]: no, fine as a ticket

File: Todo TN1.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+- implment a better way for vending the shared secret (not everyone should respond)
+
+- some restructuring of the repo (renaming of packages, consistency, etc) + better documentation, to make it easier for reviewers to understand what's gong on
+
+
+# Management contract:
+
+Implement a primitive management contract that:
+
+- stores rollups in a tree-like structure (pointing to the parent, and keeping he head)
```
[the remark this answers is not in the document]
Comment (tudor-malene) [reply Todo TN1.md:44]: better we fidn out now :)

File: Todo TN1.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
+Todo TN1
+---------
+
+#Obscuro:
+
+- evm /statedb - finalise & test properly (make sure it works through reorgs)
```
Comment (GavT) [inline Todo TN1.md:6]: This is T6

File: Todo TN1.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+- k/v on top of edgelessdb
```
Comment (GavT) [inline Todo TN1.md:8]: This is part of T4

File: Todo TN1.md (hunk 2 above)
Comment (GavT) [reply Todo TN1.md:12]: First iteration, same key every block. Next iteration would have better key derivation, one key per block

File: Todo TN1.md (hunk 1 above)
Comment (GavT) [reply Todo TN1.md:22]: Required to connect to Goerli testnet

File: Todo TN1.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (GavT) [reply Todo TN1.md:24]: Dependant on database work

File: Todo TN1.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    - reveal key after few blocks
+    - tooling to decrypt (integrated with joel's block explorer)
+    - encrypt shared secret with a key that is derived from one of the burnt in sgx keys
+
+
+- Create better logic to get L2 node up to speed. (Very primitive logic in place now, not fit for purpose if connected to ropsten)
+
+- some memory analysis - make sure a node can stay up for at least a day of rollups
+
+- rpc endpoints for everything necessary
```
Comment (GavT) [inline Todo TN1.md:26]: For now RPC endpoints being used by the simulations to call into the network. Using something like Swagger makes sense when we are exposing to users.

File: Todo TN1.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+- Create better logic to get L2 node up to speed. (Very primitive logic in place now, not fit for purpose if connected to ropsten)
+
+- some memory analysis - make sure a node can stay up for at least a day of rollups
+
+- rpc endpoints for everything necessary 
+- introduce event feeds (are these necessary for tn1?)
+
+- add configuration file support
+
+- store mapping of transaction to hash into the ethdb (see geth) - related to the statedb change
```
Comment (GavT) [inline Todo TN1.md:31]: Part of database Epic

File: Todo TN1.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+- implement better error handling - not perfect, but at least remove panics in places which are likely to blow up in a testnet scenario
```
Comment (GavT) [inline Todo TN1.md:33]: Currently not resilient against badly formed messages, so this needs better handling.

**Comment by tudor-malene** [comment]
Closing after TN1 release

**Pull request closed without merging**